[WIP - EPIC] Split Sargon into many crates

CyonAlexRDX opened the pull request:
See tracking issue https://github.com/radixdlt/sargon/issues/317

Sajjon commented:
New crates:

- sargon-core-utils
- sargon-core-error
- sargon-core-assert-json
- identified-vec-of
- sargon-core
- sargon-hierarchical-deterministic
- sargon-factors
- sargon-addresses
- sargon-keys-collector

Sajjon commented:
- sargon-core-utils
- sargon-core-error
- sargon-core-assert-json
- identified-vec-of
- sargon-core
- sargon-hierarchical-deterministic
- sargon-factors
- sargon-factors-supporting-types
- sargon-addresses
- sargon-keys-collector
- sargon-transaction-models
- sargon-manifests
- sargon-profile-supporting-types
- sargon-drivers
- http-client
- sargon-clients
- sargon-profile-gateway
- sargon-profile
- sargon-profile-app-preferences
- sargon-profile-security-structures
- sargon-profile-logic
- sargon-factor-instances-provider
- next-derivation-index-ephemeral
- gateway-models
- gateway-logic
- gateway-client-and-api

CyonAlexRDX commented:
Merge at will @GhenadieVP

Sajjon commented:
I downgraded the bump to 1.2 so that that CI check passes

Sajjon commented:
Here are some measurements:

BEFORE this PR (`main` branch):
Opening ONLY `sargon` crate
change `lib.rs` && `cargo check` => 4.50s

AFTER/WITH this PR:
change in sargon/src/lib.rs && cargo check => 1.70s

That is a 63% reduction!

CyonAlexRDX commented:
@micbakos-rdx i hope you and @GhenadieVP can coordinate to merge this PR before your changes in:

`micbakos/ABW-4027-update-io`

Branch - or any other branch. 

Merge hell otherwise 🤪

Sajjon commented:
Other changes:
I've created a type `ShortString` which **is `Copy`** 🥳, used by `DisplayName` and `Email` making them `Copy` too.

Sajjon commented:
@GhenadieVP I merge some last minute improvements: https://github.com/radixdlt/sargon/pull/327/files

I went through all `Cargo.toml` and removed all unused/unneeded dependencies. So now each crate only declares exactlty what it actually uses/needs.

An I forgot to mention in PR description:
I've removed the `Physical` variant of `FactorInstanceBadge`: https://github.com/radixdlt/sargon/blob/ac/split/crates/sargon-factors/src/factor_instance/factor_instance_badge.rs#L15-L17

We are not using it and I think we won't use it. Removing it allowed me to **not** depend on `sargon-addresses` crate in `sargon-factors` crate (as seen here https://github.com/radixdlt/sargon/blob/ac/split/crates/sargon-factors/Cargo.toml#L7-L11) and every elision of crates as dependency matters, as it builds up. Just an FYI.

OK so now I think this PR is ready to be merge from my perspective. As I wrote above merge at will :)


> [!IMPORTANT]
> I've not built Sargon locally and tried it in Android or iOS host. 
> Maybe you can do that or ask Umair to do one soundness check before mergeing?

GhenadieVP commented:
@CyonAlexRDX There is a breaking change:
```rust
pub enum AccountOrAddressOf {
    ProfileAccount { value: Account },
    AddressOfExternalAccount { value: AccountAddress },
}
```

is changed to:
```rust
pub enum OwnedOrThirdPartyAccountAddress {
    OwnedAccount { value: AccountAddress },
    ThirdPartyAccount { value: AccountAddress },
}
```

the issue is that the OwnedAccount is not an Account but just an address. I don't think that it should be the actual Account, having the AccountForDisplay form should be enough. The Host uses this information to display the info of the account, just having an address would require doing an async lookup on the profile.

micbakos-rdx commented:
> ```rust
> OwnedOrThirdPartyAccountAddress
> ```

@CyonAlexRDX No worries I have updated my changes on top of this PR.

micbakos-rdx commented:
> @micbakos-rdx i hope you and @GhenadieVP can coordinate to merge this PR before your changes in:
> 
> `micbakos/ABW-4027-update-io`
> 
> Branch - or any other branch.
> 
> Merge hell otherwise 🤪

@CyonAlexRDX No worries I have updated my changes on top of this PR.

micbakos-rdx commented:
> If anyone **sporadically** gets test failures with this (or similar) message:
> 
> > called `Result::unwrap()` on an `Err` value: FailedToDeserializeJSONToValue { json_byte_count: 0, type_name: "FactorInstancesCacheSnapshot", serde_message: "EOF while parsing a value at line 1 column 0" }
> 
> it is because crates in the same workspace **do NOT share `cfg(test)` compilation**, meaning that a non `FactorInstancesCacheClient::in_memory()` instance has been used. i.e. a real on file FactorInstancesCache is used, and two tests are running in parallell most likely causing the tests to fail.
> 
> I think I've fixed it in: [d81f018](https://github.com/radixdlt/sargon/commit/d81f018c2186d25d65935fdf9142dcfdb3d06b43)
> 
> but there might be some few more places left.

A similar test that was sometimes failing in my branch was `add_structure_emits_event`.

Sajjon commented:
> @CyonAlexRDX There is a breaking change:
> 
> ```rust
> pub enum AccountOrAddressOf {
>     ProfileAccount { value: Account },
>     AddressOfExternalAccount { value: AccountAddress },
> }
> ```
> 
> is changed to:
> 
> ```rust
> pub enum OwnedOrThirdPartyAccountAddress {
>     OwnedAccount { value: AccountAddress },
>     ThirdPartyAccount { value: AccountAddress },
> }
> ```
> 
> the issue is that the OwnedAccount is not an Account but just an address. I don't think that it should be the actual Account, having the AccountForDisplay form should be enough. The Host uses this information to display the info of the account, just having an address would require doing an async lookup on the profile.

@GhenadieVP ah right ofc, we should use `AccountForDisplay`, I will fix over the weekend!

Sajjon commented:
> A similar test that was sometimes failing in my branch was `add_structure_emits_event`.

@micbakos-rdx did your branch contain this commit: https://github.com/radixdlt/sargon/commit/2d21cc21e5b7cb9f4577886c285dc4a3b433bece

CyonAlexRDX commented:
@micbakos-rdx sorry you need to rebase again, since I merged: https://github.com/radixdlt/sargon/pull/329

@GhenadieVP i fixed the breaking change with AccountOrAddressOf and in this host PR I've fixed host: https://github.com/radixdlt/babylon-wallet-ios/pull/1429

@micbakos-rdx you probably need minor adjustments in Android host too, see https://github.com/radixdlt/babylon-wallet-ios/pull/1429

CyonAlexRDX commented:
I progress ever further with https://github.com/radixdlt/sargon/pull/330 merged

CyonAlexRDX commented:
@GhenadieVP and even further!

See https://github.com/radixdlt/sargon/pull/332

On that branch Speedups are near 10x that of main branch for common operations 

Let's sync tomorrow! I think we should merge it into this, and the this :)

Sajjon commented:
Also we should switch to only declare paths to our own crates **once**:

https://github.com/rust-lang/cargo/issues/14355#issuecomment-2573664434
